### DISCUSSION TITLE
reticulate R library for Python+R use

### DISCUSSION THREAD

**Discussion by erolbicero** [posit-dev/positron#3920] [Feedback]
Hello,

I like the IDE and its potential.  One of the powerful features of RStudio is the ability to pass variables back and forth between R and Python through the reticulate library.

Are there any plans to get this functionality in Positron?  I've tried using reticulate, and although it can run python, I can't jump back and forth between scripts like I can in RStudio.

**Comment by juliasilge** [comment]
> I can't jump back and forth between scripts like I can in RStudio.

Can you share a bit more detail about what you would like to be able to do? Maybe a simple example that demonstrates the missing functionality?

**Comment by erolbicero** [comment]
Hello,

Sure, I can provide a simple example.

If I open a fresh RStudio instance, and create two new scripts:

1. RScript.R
2. PythonScript.py


In the R script I would run:

```r
library(reticulate)

r_variable <- c(1,2,3)
```

Then I can tab over to the Python script, and run

```python
import numpy as np
import pandas as pd
```

Since I've ran `reticulate` in the first RScript, now when I go back, there should be a `py` variable _in_ R, where I can retrieve python variables, or send variables.

So I tab over to the RScript, then run
```r
py$r_variable <- r_variable
```

This should make the variable accessible in the Python script.

Now I tab back to the Python script, and run 
```python
r_variable
```

Which returns a `numpy` vector.

Still in the Python script, I can now run 
```python
py_df = pd.DataFrame(r_variable,index=['a','b','c'])
```

Now I want to access this variable in R, so I tab back to the R Script, and run
```r
py$py_df
```
which will return a `data.frame` object, without error.

It's a silly example, but in a serious workflow, it's a way to work with the same variables, transferring them back and forth between R and Python.

Please let me know if you have any other questions.

Thank you

**Reply by ntluong95** [reply]
That's very interesting. I also want to learn how to work with both Python and R. Can you share a serious workflow, like packages, tasks you utilize the two languages?

**Reply by juliasilge** [reply]
@ntluong95 You may be interested [in this blog post](https://emilhvitfeldt.com/post/pyenv-for-r-packages/) by Emil Hvitfeldt about how to set up Python envs for use in an R package.

@erolbicero Thank you for this detail. We do not currently support this kind of interactive passing of data back and forth between R and Python consoles, but it's definitely an interesting use case! I'll open an issue to track this kind of feature. (Positron does work well for rendering multilingual reticulate Quarto/`.Rmd` documents and developing packages that use reticulate.)

**Reply by juliasilge** [reply]
Here is the issue if you would like to add more info: https://github.com/posit-dev/positron/issues/3942

**Reply by erolbicero** [reply]
Thank you @juliasilge ! I really appreciate it.

@ntluong95 you could web-scrape with beautifulsoup in python, then move the data to R for statistical analysis.  Some papers have R libraries accompanying them, or Python code accompanying them.  If the code is in the "other" language, you can use that code, then analyze the results in your preferred language.  Some libraries that interact with certain databases in corporate settings are written in R or Python.  Instead of rewriting them, one can pull the data in one language, and then use the other.  I hope that's helpful.

**Comment by johol355** [comment]
I'm in the same boat as @erolbicero 

Has anyone figured out a good way to use both R and python in a Jupyter notebook? Something like this:

![](https://cdn.prod.website-files.com/654fd3ad88635290d9845b9e/6686aaaea7bd4e8421f54156_AD_4nXfNn43BAyc6RgRcrLa-nv-M4dic0tTwMTGPijsztrLtxJO3nJ9t0KmFyddlQ3ytQh2SO9hbwqcwDeyFfSOCjIVfweZ4OlaCs5MNV3UxbtxTxHRf9tv-dgQ5MwepzwI_n_9pSLlpLiDkqY2oLvgoKWkuNBSc.png)

(The example is from https://www.appsilon.com/post/introducing-positron, but I'm not convinced by the screenshot that the author got it to actually work. When I try the same thing the "R" cell can only run python code.)

**Reply by juliasilge** [reply]
I don't believe that vanilla Jupyter notebooks can use more than one kernel at the same time, without some extra Jupyter-specific extensions, and I am pretty sure that we don't support that type of workflow in Positron notebooks currently.

I create a lot of documents like this (mix of R and Python) with Quarto, so for now, I think that will be your best bet. This works because the knitr engine for Quarto can render both R and Python via reticulate. All the `.qmd` files [in this repo](https://github.com/rstudio/vetiver.rstudio.com) take this approach and I've had a good user experience developing them in Positron. I set up both an R console and Python console for interactive use, and then use the same Python env from the console to render the document via reticulate. 

(Note that we do not pass data back and forth between R and Python sessions in these examples.)

**Reply by johol355** [reply]
Thanks for the reply, Julia! I've tried Quarto with knitr and reticulate with variable success. Things start falling apart when running cells sooner or later. Interestingly, if I render the document even in the setting of cells that won't run properly, the rendered document works perfectly fine. Anyhow, it's probably as good as it gets currently.

**Comment by ZhimingYe** [comment]
In RStudio, this feature is actually implemented by `reticulate::repl_python()`, which essentially loads a line-by-line interpreter like Ipykernel within an R environment. Essentially, it relies on an R session for hosting. RStudio packages it to appear as if Python is being called directly, but in reality, it is through the reticulate package within the R environment. You can switch between R and Python within any R environment using `reticulate::repl_python()` and `quit/exit` command in any terminal with an opened R.

I think adding this feature to Positron would be challenging because it might be unclear initially whether the user needs a pure-Python environment or a Python environment within `reticulate::repl_python()`. Sometimes, the warpper of reticulate might not be satisfactory, which is one of the reasons why Python support in RStudio is criticized. I am curious to see how the Positron team will address this issue. My suggestion is to provide an option for reticulate/ipykernel when creating a project, but I am not sure if this is the best approach. Once again, thank you to the Positron team for your outstanding work; I am increasingly interested in this IDE! :)

<img width="1003" alt="image" src="https://github.com/posit-dev/positron/assets/96439664/da26d81b-d9b5-488b-bd37-b10d3ceabf61">

**Reply by erolbicero** [reply]
I would be okay with the choice between vanilla Python, and then an R-Repl version.  Most Python users wouldn't use it, and then a subset of R users presumably would.

**Comment by p0bs** [comment]
FWIW, I think this be the truly 'killer functionality' in positron/quarto if it were possible. After all, to support multi-lingual teams, it would be great if someone's python code were augmented by another person's R code, with work continuing thereafter (in either language). The passing of variables between R and python code chunks (with environments managed) sounds really challenging from a technical perspective but would be perfect for this user.

To be clear, I'm agnostic on how you might achieve this as the functionality is key (and I trust your design choices more than mine).

Continued thanks for all your great work.

**Comment by dfalbel** [comment]
Hi everyone, 

We have added support for reticulate console sessions in #4603. It's currently turned off by default, but can be enabled by toggling the `positron.reticulate.enabled` setting (Open user preferences and search for reticulate). You'll also need a recent Positron build, eg the latest Positron Version: 2024.09.0 (Universal) build 103.

We'd love to hear your feedback on this feature.

Thank you very much!

**Reply by KGZaker** [reply]
Hi, thank you so much for sharing this exciting news! However, I’ve noticed that while I can now access the Reticulate Python variable in R (i.e., `py$tmp_var` in R), I’m unable to do the reverse — meaning I can’t access an R variable in Reticulate Python (i.e., `r.tmp_var` in Python). I was wondering if this is an issue specific to my setup, or if this functionality is not currently supported.

**Reply by dfalbel** [reply]
Hi, we're working on an issue that doesn't allow it to be visualized from the variables pane by default. But with the dev version of reticulate you should be able to do `r.tmp_var` from Python too. It has been fixed in https://github.com/rstudio/reticulate/pull/1681
